File: mcp/kotlin-mcp-server/server/src/main/kotlin/com/yourorg/mcp/schema/Validator.kt
package com.yourorg.mcp.schema

import com.networknt.schema.JsonSchema
import com.networknt.schema.JsonSchemaFactory
import com.networknt.schema.SpecVersion
import com.networknt.schema.ValidationMessage
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.encodeToJsonElement

// Custom exceptions
class ValidationException(
    val code: String,
    val details: String,
    val errors: List<ValidationMessage> = emptyList()
) : Exception("$code: $details")

class SchemaMissingException(toolName: String) :
    Exception("Schema not found for tool: $toolName")

// JSON Schema factory with strict mode
private val jsonSchemaFactory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V202012)

// JSON configuration with strict mode
private val json = Json {
    encodeDefaults = true
    explicitNulls = false
    ignoreUnknownKeys = false
    isLenient = false
    allowStructuredMapKeys = true
    prettyPrint = false
    useArrayPolymorphism = false
}

// Schema storage - in production, this would be loaded from the schema file
private val toolSchemas = mutableMapOf<String, Pair<JsonSchema, JsonSchema>>()

// Initialize schemas (this would be done at startup)
fun initializeSchemas(schemaJson: String) {
    // Parse the main schema file and extract tool schemas
    // This is a simplified implementation - in production you'd parse the full schema
    val schemaMap = mapOf(
        "refactorFunction" to createToolSchema(
            """
            {
              "type": "object",
              "required": ["filePath", "functionName", "refactorType"],
              "properties": {
                "filePath": {"type": "string", "minLength": 1},
                "functionName": {"type": "string", "minLength": 1},
                "refactorType": {"type": "string", "enum": ["rename", "extract", "inline", "introduceParam"]},
                "newName": {"type": "string"},
                "preview": {"type": "boolean", "default": false}
              }
            }
            """,
            """
            {
              "type": "object",
              "required": ["patch", "affectedFiles"],
              "properties": {
                "patch": {"type": "string", "minLength": 1},
                "affectedFiles": {"type": "array", "items": {"type": "string", "minLength": 1}}
              }
            }
            """
        ),
        // Add other tool schemas here...
    )

    toolSchemas.putAll(schemaMap)
}

private fun createToolSchema(inputSchema: String, outputSchema: String): Pair<JsonSchema, JsonSchema> {
    val inputJsonSchema = jsonSchemaFactory.getSchema(inputSchema)
    val outputJsonSchema = jsonSchemaFactory.getSchema(outputSchema)
    return Pair(inputJsonSchema, outputJsonSchema)
}

// Validation functions
fun validateInput(toolName: String, jsonElement: JsonElement): Unit {
    val schemas = toolSchemas[toolName] ?: throw SchemaMissingException(toolName)
    val inputSchema = schemas.first

    val errors = inputSchema.validate(jsonElement)
    if (errors.isNotEmpty()) {
        throw ValidationException(
            code = "ValidationError",
            details = "Input validation failed for $toolName",
            errors = errors
        )
    }
}

fun validateOutput(toolName: String, jsonElement: JsonElement): Unit {
    val schemas = toolSchemas[toolName] ?: throw SchemaMissingException(toolName)
    val outputSchema = schemas.second

    val errors = outputSchema.validate(jsonElement)
    if (errors.isNotEmpty()) {
        throw ValidationException(
            code = "InternalOutputInvalid",
            details = "Output validation failed for $toolName",
            errors = errors
        )
    }
}

// Convenience functions for data classes
inline fun <reified T> validateInput(toolName: String, data: T): Unit {
    val jsonElement = json.encodeToJsonElement(data)
    validateInput(toolName, jsonElement)
}

inline fun <reified T> validateOutput(toolName: String, data: T): Unit {
    val jsonElement = json.encodeToJsonElement(data)
    validateOutput(toolName, jsonElement)
}

// Utility functions
fun getAvailableTools(): Set<String> = toolSchemas.keys

fun hasTool(toolName: String): Boolean = toolSchemas.containsKey(toolName)

// Schema loading from file (production implementation)
fun loadSchemasFromFile(schemaFilePath: String) {
    // Implementation would parse the JSON schema file and populate toolSchemas
    // For now, this is a placeholder
}

File: mcp/kotlin-mcp-server/server/src/schema/validator.ts
/**
 * MCP Tool Validators - TypeScript Implementation
 *
 * Uses Ajv v8 (2020-12) with ajv-formats for schema validation.
 */

import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import { toolSchemas } from './types';

// Initialize Ajv with 2020-12 draft support
const ajv = new Ajv({
  strict: true,
  allErrors: true,
  useDefaults: true,
  coerceTypes: false,
  removeAdditional: false,
  validateFormats: true,
  $data: true,
});

// Add format support
addFormats(ajv);

// Compile validators for each tool
const inputValidators: { [toolName: string]: Ajv.ValidateFunction } = {};
const outputValidators: { [toolName: string]: Ajv.ValidateFunction } = {};

for (const [toolName, schemas] of Object.entries(toolSchemas)) {
  try {
    inputValidators[toolName] = ajv.compile(schemas.input);
    outputValidators[toolName] = ajv.compile(schemas.output);
  } catch (error) {
    console.error(`Failed to compile validator for ${toolName}:`, error);
  }
}

// Custom error classes
export class ValidationError extends Error {
  constructor(
    public code: string,
    public message: string,
    public data?: any
  ) {
    super(message);
    this.name = 'ValidationError';
  }
}

export class SchemaMissingError extends Error {
  constructor(toolName: string) {
    super(`Schema not found for tool: ${toolName}`);
    this.name = 'SchemaMissingError';
  }
}

// Validation functions
export function validateInput(toolName: string, data: any): void {
  const validator = inputValidators[toolName];
  if (!validator) {
    throw new SchemaMissingError(toolName);
  }

  const valid = validator(data);
  if (!valid) {
    throw new ValidationError(
      'ValidationError',
      `Input validation failed for ${toolName}`,
      { errors: validator.errors }
    );
  }
}

export function validateOutput(toolName: string, data: any): void {
  const validator = outputValidators[toolName];
  if (!validator) {
    throw new SchemaMissingError(toolName);
  }

  const valid = validator(data);
  if (!valid) {
    throw new ValidationError(
      'InternalOutputInvalid',
      `Output validation failed for ${toolName}`,
      { errors: validator.errors }
    );
  }
}

// Utility function to get available tools
export function getAvailableTools(): string[] {
  return Object.keys(toolSchemas);
}

// Utility function to get tool schema
export function getToolSchema(toolName: string) {
  return toolSchemas[toolName];
}
